pkg/object/graph/depends.go

```diff
@@ -80,73 +96,140 @@ func ReverseSortObjs(objs object.UnstructuredSet) ([]object.UnstructuredSet, err
 	return s, nil
 }
 
-// addApplyTimeMutationEdges updates the graph with edges from objects
-// with an explicit "apply-time-mutation" annotation.
-func addApplyTimeMutationEdges(g *Graph, objs object.UnstructuredSet) {
-	for _, obj := range objs {
-		id := object.UnstructuredToObjMetadata(obj)
+// addVertices adds all the IDs in the set as graph vertices.
+func addVertices(g *Graph, ids object.ObjMetadataSet) {
+	for _, id := range ids {
 		klog.V(3).Infof("adding vertex: %s", id)
 		g.AddVertex(id)
-		if mutation.HasAnnotation(obj) {
-			subs, err := mutation.ReadAnnotation(obj)
-			if err != nil {
-				// TODO: fail task if parse errors?
-				klog.V(3).Infof("failed to add edges from: %s: %s", id, err)
-				return
+	}
+}
+
+// addApplyTimeMutationEdges updates the graph with edges from objects
+// with an explicit "apply-time-mutation" annotation.
+// The objs and ids must match in order and length (optimization).
```

Comment:
Is this to avoid having to translate from unstructured to ObjMetadata many times?

Reply:
Yes. It's just a minor optimization and only affects private functions.

I considered passing a map of id->obj, but then you lose the ordering, which makes testing painful or requires sorting again, which is less efficient.

That said, the sets are all being sorted at the end of SortObjs, so the end result would be mostly the same. The gotcha is that if the sort only sorts by type, then the output will be different if the input is different (ex: two objects of the same type in a different order).